bindings/cpu-time.cc

```diff
@@ -0,0 +1,68 @@
+#include "cpu-time.hh"
+
+#ifdef __linux__
+#include <errno.h>
+#include <pthread.h>
+#include <string.h>
+#elif __APPLE__
+#define _DARWIN_C_SOURCE
+#include <mach/mach_error.h>
+#include <mach/mach_init.h>
+#include <mach/thread_act.h>
+#endif
+
+namespace dd {
+
+CpuTime::CpuTime(struct timespec time) : last_(time) {
+#ifdef __linux__
+  pthread_getcpuclockid(pthread_self(), &clockid);
+#elif __APPLE__
+  thread_ = mach_thread_self();
+#endif
+}
+
+CpuTime::CpuTime()
+  : CpuTime(Now()) {}
+
+int64_t CpuTime::Diff(struct timespec now) {
+  int64_t current = now.tv_sec * INT64_C(1000000000) + now.tv_nsec;
+  int64_t prev = last_.tv_sec * INT64_C(1000000000) + last_.tv_nsec;
+  int64_t cpu_time = current - prev;
+
+  last_ = now;
+
+  return cpu_time;
+}
+
+int64_t CpuTime::Diff() {
+  return Diff(Now());
+}
+
+struct timespec CpuTime::Now() {
+  struct timespec cpu_time = {.tv_sec = 0, .tv_nsec = 0};
+
+#ifdef __linux__
+  if (clock_gettime(clockid, &cpu_time)) {
+    return (struct timespec){.tv_sec = 0, .tv_nsec = 0};
+  }
+#elif __APPLE__
+  mach_msg_type_number_t count = THREAD_BASIC_INFO_COUNT;
```

Comment:
TIL I did not know that APIs were so different in MacOS. Did you already run some tests there ?

Reply:
Yep, one of the first parts I wrote tests for because it was an easy low-effort place to try out the C++ testing strategy I was working on. That code is also basically copy/pasted from the PHP profiler, so should be correct as far as I'm aware.